heat_pump_adoption_modelling/pipeline/MCS/mcs_epc_joining.py

```diff
@@ -0,0 +1,322 @@
+    comp.string(
+        "standardised_address",
+        "standardised_address",
+        method="jarowinkler",
+        label="address_score",
+    )
+
+    # Classify
+    print("- Computing a matching...")
+    matching = comp.compute(candidate_pairs, df1, df2)
+
+    return matching
+
+
+def join_mcs_epc_data(dhps=None, epcs=None, save=True, drop_epc_address=False):
+    """Joins MCS and EPC data.
+
+    Parameters
+    ----------
+    dhps : pandas.Dataframe
+        Dataframe with standardised_postcode, numeric_tokens
+        and standardised_address fields.
+        If None, domestic HP data is loaded and augmented.
+    epcs : pandas.Dataframe
+        Dataframe with standardised_postcode, numeric_tokens
+        and standardised_address fields.
+        If None, EPC data is loaded and augmented.
+    save : bool
+        Whether or not to save the output.
+    drop_epc_address : bool
+        Whether or not to drop addresses from the EPC records.
+        Useful to keep for determining whether matches are sensible.
+
+    Return
+    ----------
+    merged : pandas.Dataframe
+        Dataframe containing merged MCS and EPC records."""
+
+    if dhps is None:
+        print("Preparing domestic HP data...")
+        dhps = load_domestic_hps()
+        dhps = prepare_dhps(dhps)
+
+    if epcs is None:
+        print("Preparing EPC data...")
+        epcs = load_epcs()
+        epcs = prepare_epcs(epcs)
+
+    print("Forming a matching...")
+    matching = form_matching(df1=dhps, df2=epcs)
+
+    # First ensure that all matches are above the matching parameter
+    good_matches = matching[
+        (matching["numerics"] == 1) & (matching["address_score"] >= matching_parameter)
+    ].reset_index()
+
+    # Take the indices of the rows with best address match for each MCS index
+    top_match_indices = good_matches.groupby("level_0")["address_score"].idxmax()
+
+    # Filter the matches df to just the top matches
+    # Drop duplicates in case of ties
+    top_matches = (
+        good_matches.loc[top_match_indices]
+        .drop_duplicates("level_0")
+        .drop(columns=["numerics", "address_score"])
+    )
+
+    print("Joining the data...")
+    merged = (
+        dhps.reset_index()
+        .merge(top_matches, how="left", left_on="index", right_on="level_0")
+        .merge(epcs.reset_index(), how="left", left_on="level_1", right_on="index")
+        .drop(
+            columns=[
+                "index_x",
+                "standardised_postcode_x",
+                "standardised_address_x",
+                "numeric_tokens_x",
+                "ADDRESS1",
+                "POSTTOWN",
+                "POSTCODE",
+                "index_y",
+                "standardised_postcode_y",
+                "numeric_tokens_y",
+            ]
+        )
+    )
+
+    if drop_epc_address:
+        merged = merged.drop(columns="standardised_address_y")
+    else:
+        merged = merged.rename({"standardised_address_y": "epc_address"})
```

Comment:
This doesn't quite work yet, you'll need a keyword.

```suggestion
        merged = merged.rename(columns={"standardised_address_y": "epc_address"})
```

Reply:
Ah yeah, thanks!